File: 46-permutations/permutations.cpp
class Solution {
private:
    void solve(vector<int> &nums, unordered_set<int> &st, vector<vector<int>> &ans, vector<int> &temp, int idx){
        if (idx == nums.size()){
            ans.push_back(temp);
            return;
        }

        for(auto &it:nums){
            if (st.count(it) == 0){
                st.insert(it);
                temp.push_back(it);
                solve(nums, st, ans, temp, idx+1);
                st.erase(it);
                temp.pop_back();
            }
        }
    }
public:
    vector<vector<int>> permute(vector<int>& nums) {
        unordered_set<int> st;
        vector<int> temp;
        vector<vector<int>> ans;

        solve(nums, st, ans, temp, 0);

        return ans;
    }
};

/*
    - We need to find all the permutation of the given array nums
    - The length of nums is <= 8, so we can go for backtracking kind of solution. 
    - How do we use the backtracking?
        -
*/

File: 46-permutations/permutations.go
func solve(nums []int, idx int, temp []int, ans *[][]int, set map[int]int) {
    if idx == len(nums) {
        tmp := make([]int, len(temp))
        copy(tmp, temp)
        *ans = append(*ans, tmp)

        return
    }

    for i := range nums {
        if _, ok := set[nums[i]]; !ok {
            temp = append(temp, nums[i])
            set[nums[i]] = 1
            solve(nums, idx+1, temp, ans, set)
            delete(set, nums[i])
            temp = temp[:len(temp)-1]
        }
    }
}

func permute(nums []int) [][]int {
    var temp []int
    var ans [][]int
    set := make(map[int]int)
    solve(nums, 0, temp, &ans, set)

    return ans
}